File: boj/cpp/2820.cpp
#include <bits/stdc++.h>
#include <iostream>
using namespace std;

typedef pair<int, int> Query;

int n, m;
vector<Query> queries;
vector<int> graph[500001];
vector<long> arr;
vector<long> tree;
vector<long> lazy;

void propagate(int node, int start, int end)
{
    if (lazy[node])
    {
        if (start != end)
        {
            lazy[node << 1] += lazy[node];
            lazy[node << 1 | 1] += lazy[node];
        }
        
        tree[node] += (end - start + 1) * lazy[node];
        lazy[node] = 0;
    }
}

void update(int left, int right, long value, int node=1, int start=1, int end=n)
{
    propagate(node, start, end);
    
    if (end < left || right < start)
        return;
    
    if (left <= start && end <= right)
    {
        lazy[node] += value;
        propagate(node, start, end);
        return;
    }
    
    int mid = (start + end) >> 1;
    update(left, right, value, node << 1, start, mid);
    update(left, right, value, node << 1 | 1, mid + 1, end);
    
    tree[node] = tree[node << 1] + tree[node << 1 | 1];
}

long query(int index, int node=1, int start=1, int end=n)
{
    propagate(node, start, end);
    
    if (start == end)
        return tree[node];
        
    int mid = (start + end) >> 1;
    if (index > mid)
        return query(index, node << 1 | 1, mid + 1, end);
    else
        return query(index, node << 1, start, mid);
}

int euler(int node, int start)
{
    queries[node].first = start;
    for(auto x : graph[node])
    {
        if (!queries[x].first)
            start = euler(x, start + 1);
    }
    queries[node].second = start;
    return start;
}

int main()
{
    cin.tie(0); cout.tie(0);
    ios_base::sync_with_stdio(false);
    
    cin >> n >> m;
    arr.resize(n + 1, 0);
    queries.resize(n + 1, {0, 0});
    tree.reserve(4 * n);
    lazy.reserve(4 * n);
    
    cin >> arr[1];
    for(int i=2; i<=n; i++)
    {
        int parent;
        cin >> arr[i] >> parent;
        graph[parent].push_back(i);
    }
    
    euler(1, 1);
    
    for(int i=1; i<=n; i++)
    {
        int index = queries[i].first;
        update(index, index, arr[i]);
    }
    
    while(m--)
    {
        char q; cin >> q;
        
        if (q == 'p')
        {
            int index;
            long value;
            cin >> index >> value;
            
            int left = queries[index].first + 1;
            int right = queries[index].second;
            
            if (left > right)
                continue;
            
            update(left, right, value);
        }
        else
        {
            int index; cin >> index;
            cout << query(queries[index].first) << '\n';
        }
    }

    return 0;
}

File: boj/python/Segment Tree/Lazy propagation/2820.py
import sys
import math
input = sys.stdin.readline

def euler(tree, n):
    index = 0
    stack = [1]
    start = [0] * (n + 1)
    end = [0] * (n + 1)
    
    while stack:
        node = stack[-1]
        
        if not start[node]:
            index += 1
            start[node] = index
            
            for next in tree[node]:
                stack.append(next)
        else:
            end[node] = index
            stack.pop()
    
    return start, end

def solve():
    n, m = map(int, input().split())
    
    salary = [0] * (n + 1)
    graph = [[] for _ in range(n + 1)]
    
    salary[1] = int(input())
    for x in range(2, n + 1):
        value, parent = map(int, input().split())
        salary[x] = value
        graph[parent].append(x)
    
    starts, ends = euler(graph, n)
    
    height = math.ceil(math.log2(n))
    size = 1 << (height + 1)
    tree = [0 for _ in range(size)]
    lazy = [0 for _ in range(size)]
    
    def propagate(node, start, end):
        if lazy[node]:
            if start != end:
                lazy[node << 1] += lazy[node]
                lazy[node << 1 | 1] += lazy[node]
            
            tree[node] += (end - start + 1) * lazy[node]
            lazy[node] = 0
    
    def update(left, right, value, node=1, start=1, end=n):
        propagate(node, start, end)
        
        if end < left or right < start:
            return
        
        if left <= start and end <= right:
            lazy[node] += value
            propagate(node, start, end)
            return

        mid = (start + end) >> 1
        update(left, right, value, node << 1, start, mid)
        update(left, right, value, node << 1 | 1, mid + 1, end)
        
        tree[node] = tree[node << 1] + tree[node << 1 | 1]
        
    def query(index, node=1, start=1, end=n):
        propagate(node, start, end)
        
        if start == end:
            return tree[node]
        
        mid = (start + end) >> 1
        if index <= mid:
            return query(index, node << 1, start, mid)
        else:
            return query(index, node << 1 | 1, mid + 1, end)
        
    for x in range(1, n + 1):
        update(starts[x], starts[x], salary[x])
    
    for _ in range(m):
        q, *a = input().strip().split()
        
        if q == 'p':
            index, value = map(int, a)
            if starts[index] > ends[index]:
                continue
            
            update(starts[index] + 1, ends[index], value)
        else:
            index, = map(int, a)
            print(query(starts[index]))
    
solve()